k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt

```diff
@@ -57,14 +59,26 @@ class K8sResourceHandler (
         resource: Resource<K8sResourceSpec>,
         diff: ResourceDiff<K8sObjectManifest>
     ): List<Task> {
-
         // send a notification on attempt to deploy the image artifact
-        val imageString = find(resource.spec.template.spec as MutableMap<String, Any?>, "image") as String?
-        imageString?.let{
-            log.info("Deploying artifact $it")
-            notifyArtifactDeploying(resource, getTag(it))
+        if (!resource.spec.template.spec.isNullOrEmpty()) {
+            val imageString = find(resource.spec.template.spec as MutableMap<String, Any?>, "image") as String?
+            imageString?.let{
+                log.info("Deploying artifact $it")
+                notifyArtifactDeploying(resource, getTag(it))
+            }
+        }
+        // Spinnaker versions configMap and secrets by default
+        if (!resource.spec.template.data.isNullOrEmpty()) {
+            val copy =  resource.spec.template.metadata.toMutableMap()
+            if (copy.containsKey("annotations")) {
+                val copyMap = (copy["annotations"] as Map<String, Any?>).toMutableMap()
+                copyMap["strategy.spinnaker.io/versioned"] = "false"
+                copy["annotations"] = copyMap
+            } else {
+                copy["annotations"] = mapOf("strategy.spinnaker.io/versioned" to "false")
+            }
+            resource.spec.template.metadata = copy
         }
```

Comment:
unclear about this bit. We ask for the annotation to be removed here only to put it back in this code snippet. what's the reason?

Reply:
When you are asking Clouddriver to deploy manifests, clouddriver will add the `strategy.spinnaker.io/versioned` annotation and version configMaps and secrets. If the annotation is present and set to false, it will not do that. 

During upsert, this annotation is added for the reason above. When you get the manifest from kubernetes for this resource during resource check, the annotation field contains the `strategy.spinnaker.io/versioned` key. When you do a diff against this and resource spec without removing the key, keel will never realize it's in sync because technically it's different. (The `last-applied-configuration` contains the annotation key while resource spec in keel doesn't)

Reply:
Thinking about this, wonder if upsert is not the right place to do this. Upsert probably should not touch manifests being deployed. Should be done in `toResolvedType`?

Reply:
I think I came across this when I was writing the `cleanup` code but for some reason it wasn't a problem then. Do you remember why?

what do we lose if we add the `strategy.spinnaker.io/versioned: false` annotation to a resource for which clouddriver has had a versioning strategy? wouldn't it result in clouddriver not picking up changes and hence not reconciling things properly?

To your point, I think if we do it in `toResolvedType` we can modify desired, so we dont need to remove once and add it back again in two different places.

Reply:
Spinnaker adds that particular annotation for configmap and secrets only. This plugin hasn't worked for these resource types so it wasn't a problem. But now we are supporting these types, we are seeing this issue. 
If we do not annotate, the configMap looks something like this for a resource named `game-config`
```
NAME               DATA   AGE
game-config-v000   2      4d20h
```
This is fine as it is. Problem is with current interface for clouddriver k8s provider, there's no way to get this resource and keel will think the resource does not exist and tries to provision endlessly. 
The right thing to do is to find a way to get versioned resources through clouddriver and get the correct version. 
This is what I meant by:
```
ConfigMaps created through clouddriver is versioned by default. We add the strategy.spinnaker.io/versioned annotation to disable it for now. We may need to revisit this in the future to take advantage of versioned secrets and configmaps.
```
This is out of scope for this PR IMO

Reply:
yup sounds good.